src/pages/SupplierInvoicePage.js

```diff
@@ -1,398 +1,328 @@
+  DisabledUncheckedComponent,
+  NewExpiryDateInput,
 } from '../widgets';
-import { GenericPage } from './GenericPage';
-import { BottomConfirmModal, PageContentModal } from '../widgets/modals';
-
-import { createRecord } from '../database';
-import { formatDate, parsePositiveInteger, sortDataBy } from '../utilities';
-import { buttonStrings, modalStrings, pageInfoStrings, tableStrings } from '../localization';
-
-import globalStyles, { dataTableStyles } from '../globalStyles';
-
-const DATA_TYPES_SYNCHRONISED = ['TransactionItem', 'TransactionBatch', 'Item', 'ItemBatch'];
-
-const MODAL_KEYS = {
-  COMMENT_EDIT: 'commentEdit',
-  THEIR_REF_EDIT: 'theirRefEdit',
-  ITEM_SELECT: 'itemSelect',
-};
-
-const SORT_DATA_TYPES = {
-  itemName: 'string',
-  itemCode: 'string',
-  batch: 'string',
-  totalQuantity: 'number',
-  packSize: 'number',
-};
-
-export class SupplierInvoicePage extends React.Component {
-  constructor(props) {
-    super(props);
-    this.dataFilters = {
-      searchKey: '',
-      sortBy: 'itemName',
-      isAscending: true,
-    };
-    this.state = {
-      modalKey: null,
-      modalIsOpen: false,
-      selection: [],
+import {
+  DataTable,
+  Row,
+  Cell,
+  EditableCell,
+  CheckableCell,
+  DataTableHeaderRow,
+} from '../widgets/DataTable';
+import {
+  focusCell,
+  focusNext,
+  selectRow,
+  deselectRow,
+  deselectAll,
+  sortData,
+  filterData,
+  closeBasicModal,
+  editComment,
+  editTheirRef,
+  openBasicModal,
+  refreshData,
+  editTransactionBatchQuantity,
+  deleteTransactionBatchesById,
+  editTransactionBatchExpiryDate,
+  addTransactionBatch,
+} from './dataTableUtilities/actions';
+
+import globalStyles, { SUSSOL_ORANGE, newDataTableStyles, newPageStyles } from '../globalStyles';
+import usePageReducer from '../hooks/usePageReducer';
+import DataTablePageView from './containers/DataTablePageView';
+
+const keyExtractor = item => item.id;
+
+export const SupplierInvoicePage = ({ routeName, transaction }) => {
+  const [state, dispatch, instantDebouncedDispatch, debouncedDispatch] = usePageReducer(routeName, {
+    pageObject: transaction,
+    backingData: transaction.getTransactionBatches(UIDatabase),
+    data: transaction
+      .getTransactionBatches(UIDatabase)
+      .sorted('itemName')
+      .slice(),
+    keyExtractor,
+    dataState: new Map(),
+    currentFocusedRowKey: null,
+    searchTerm: '',
+    filterDataKeys: ['itemName'],
+    sortBy: 'itemName',
+    isAscending: true,
+    modalKey: '',
+    hasSelection: false,
+  });
+
+  const {
+    data,
+    dataState,
+    sortBy,
+    isAscending,
+    columns,
+    modalKey,
+    pageInfo,
+    pageObject,
+    hasSelection,
+    backingData,
+  } = state;
+
+  const { isFinalised, comment, theirRef } = pageObject;
+  const { ITEM_SELECT, COMMENT_EDIT, THEIR_REF_EDIT } = MODAL_KEYS;
+
+  if (isFinalised && data.length !== backingData.length) dispatch(refreshData());
+
+  const getItemLayout = useCallback((_, index) => {
```

Comment:
extracting this could be a good idea, every page going to use it, I'd say. Maybe this and `keyExtractor` could be put into maube `dataTable/utilities`

Reply:
Agree